Orange/data/table/base.py

```diff
@@ -0,0 +1,1242 @@
+            if len(weight) != len(self):
+                raise ValueError("The sequence has length {}, expected length {}.".format(len(weight), len(self)))
+            new_weights = weight
+        elif isinstance(weight, np.ndarray):  # np.ndarray is not a Sequence
+            # allow row or column vectors
+            if weight.ndim > 1 and not (weight.ndim == 2 and weight.shape[1] == 1):
+                raise ValueError("Dimension mismatch.")
+            new_weights = np.ravel(weight)
+            if len(weight) != len(self):
+                raise ValueError("There are {} weights, expected {}.".format(len(weight), len(self)))
+        elif isinstance(weight, pd.Series):  # not only SeriesBase
+            # drop everything but the values to uncomplicate things
+            if weight.isnull().any():
+                raise ValueError("Weights cannot be nan. ")
+            new_weights = weight.values
+        elif isinstance(weight, pd.Categorical):
+            new_weights = list(weight)
+        else:
+            raise TypeError("Expected one of [Number, str, Sequence, SeriesBase].")
+
+        # final check that we're actually adding numbers as weights
+        if not (isinstance(new_weights, Number) or all(isinstance(w, Number) for w in new_weights)):
+            raise ValueError("All weight values must be numbers.")
+        self[self._WEIGHTS_COLUMN] = new_weights
+
+    @property
+    def has_weights(self):
+        """Check if this Table has any weights.
+
+        Weight presence is determined by all weights not being identical.
+
+        Returns
+        -------
+        bool
+            True if the weights are not all identical, False otherwise.
+        """
+        return len(self) and len(self[self._WEIGHTS_COLUMN].unique()) != 1
+
+    @classmethod
+    def _new_id(cls, num=1, force_list=False):
+        """Generate new application-wide globally-unique numbers.
+
+        Parameters
+        ----------
+        num : int, optional, default 1
+            The number of new IDs to generate.
+        force_list : bool, optional, default False
+            Whether to always force the result to be a list,
+            even if only one number is generated.
+
+        Returns
+        -------
+        int or list[int]
+            New globally-unique identifiers.
+            If generating a single number and not forcing into a list, output an integer.
+            Otherwise output a list of integers of len(num).
+        """
+        with cls._next_instance_lock:
+            out = np.arange(cls._next_instance_id, cls._next_instance_id + num)
+            cls._next_instance_id += num
+            return out[0] if num == 1 and not force_list else out
+
+    def _transform_discrete_values_to_descriptors(self):
+        """Transform discrete variables given in index form into descriptors.
+
+        The conversion is performed in-place.
+        """
+        for var in chain(self.domain.variables, self.domain.metas):
```

Comment:
See, this. This is so fucking idiotic. Search the codebase for `chain()`.  This is everywhere.